src/main/java/teammates/storage/api/StudentsDb.java

```diff
@@ -403,31 +404,43 @@ public void updateStudent(String courseId, String email, String newName,
         CourseStudent courseStudent = getCourseStudentEntityForEmail(courseId, email);
         if (courseStudent != null) {
             boolean isEmailChanged = !email.equals(newEmail);
+            String lastName = StringHelper.splitName(newName)[1];
+            
             if (isEmailChanged) {
-                CourseStudent courseStudentWithNewEmail = getCourseStudentEntityForEmail(courseId, newEmail);
-                if (courseStudentWithNewEmail != null) {
+                CourseStudent newCourseStudent = new CourseStudent(newEmail, newName, newGoogleId, newComments,
+                                                                   courseId, newTeamName, newSectionName);
+                newCourseStudent.setLastName(lastName);
+                newCourseStudent.setCreatedAt(courseStudent.getCreatedAt());
+                if (keepUpdateTimestamp) {
+                    newCourseStudent.setLastUpdate(courseStudent.getUpdatedAt());
+                }
+                
+                try {
+                    createStudent(new StudentAttributes(newCourseStudent), hasDocument);
+                } catch (EntityAlreadyExistsException e) {
+                    StudentAttributes existingStudent = (StudentAttributes) e.existingEntity;
                     String error = ERROR_UPDATE_EMAIL_ALREADY_USED
-                            + courseStudentWithNewEmail.getName() + "/" + courseStudentWithNewEmail.getEmail();
+                            + existingStudent.getName() + "/" + existingStudent.getEmail();
                     throw new InvalidParametersException(error);
                 }
-            }
-    
-            courseStudent.setEmail(newEmail);
-            courseStudent.setName(newName);
-            courseStudent.setLastName(StringHelper.splitName(newName)[1]);
-            courseStudent.setComments(newComments);
-            courseStudent.setGoogleId(newGoogleId);
-            courseStudent.setTeamName(newTeamName);
-            courseStudent.setSectionName(newSectionName);
+                
+                deleteStudent(courseId, email);
```

Comment:
Not related to this PR: this can be a cause of errors if the request was terminated before the student was deleted. We should investigate use of datastore level transactions to ensure either both operations happen or both fail.